pkg/controller/docdb/dbcluster/dbcluster.go

```diff
@@ -0,0 +1,304 @@
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/docdb/v1alpha1"
+
+	"context"
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	errNotDBCluster     = "managed resource is not a DocDB DBCluster custom resource"
+	errKubeUpdateFailed = "cannot update DocDBCluster custom resource"
+)
+
+// SetupDBCluster adds a controller that reconciles a DBCluster.
+func SetupDBCluster(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.DBClusterKind)
+	opts := []option{setupExternal}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&svcapitypes.DBCluster{}).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.DBClusterGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), managed.NewNameAsExternalName(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func setupExternal(e *external) {
+	e.preObserve = preObserve
+	e.postObserve = postObserve
+	h := &hooks{client: e.client, kube: e.kube}
+	e.isUpToDate = h.isUpToDate
+	e.preUpdate = h.preUpdate
+	e.postUpdate = h.postUpdate
+	e.preCreate = h.preCreate
+	e.postCreate = h.postCreate
+	e.preDelete = preDelete
+	e.filterList = filterList
+}
+
+type hooks struct {
+	client docdbiface.DocDBAPI
+	kube   client.Client
+}
+
+type tagger struct {
+	kube client.Client
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.DescribeDBClustersInput) error {
+	obj.DBClusterIdentifier = awsclient.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.DBCluster, resp *svcsdk.DescribeDBClustersOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	switch awsclient.StringValue(cr.Status.AtProvider.Status) {
+	case svcapitypes.DocDBInstanceStateAvailable:
+		cr.Status.SetConditions(xpv1.Available())
+	case svcapitypes.DocDBInstanceStateCreating:
+		cr.Status.SetConditions(xpv1.Creating())
+	case svcapitypes.DocDBInstanceStateDeleting:
+		cr.Status.SetConditions(xpv1.Deleting())
+	default:
+		cr.Status.SetConditions(xpv1.Unavailable())
+	}
+
+	return obs, nil
+}
+
+func (e *hooks) isUpToDate(cr *svcapitypes.DBCluster, resp *svcsdk.DescribeDBClustersOutput) (bool, error) { // nolint:gocyclo
+	cluster := resp.DBClusters[0]
+
+	if (cr.Spec.ForProvider.BackupRetentionPeriod != nil && !awsclient.IsEqualInt64(cr.Spec.ForProvider.BackupRetentionPeriod, cluster.BackupRetentionPeriod)) ||
```

Comment:
```suggestion
	if awsclient.Int64Value(cr.Spec.ForProvider.BackupRetentionPeriod) == awsclient.Int64Value(cluster.BackupRetentionPeriod) ||
```

It should take care of `nil` possibility as well. Goes for other comparisons as well.

Reply:
Also, I'd suggest using `switch {}` as it's easier to read when there are many `OR` operators.

Reply:
Fixed